### PR TITLE
(GRAM, PSI, FMT): introduce foreign_mod_body and mod_body

### PR DESCRIPTION
**Pull request by jajakobyly**

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/grammar/rust.bnf [hunk 1]
```diff
@@ -526,11 +526,17 @@ extern_crate_item ::= EXTERN CRATE IDENTIFIER alias? ';'
 
 mod_decl_item ::= MOD IDENTIFIER ';'
 
-mod_item ::= MOD IDENTIFIER '{' inner_attr* docs_and_item* '}' {
-  pin(".*") = 3
+mod_item ::= MOD IDENTIFIER mod_body {
+  methods = [
+    // FIXME(jajakobyly): This method is tagged as @Nullable while we want @NotNull
+    itemList = "/mod_body/item_"
+  ]
```
Comment (jajakobyly) [inline src/main/kotlin/org/rust/lang/core/grammar/rust.bnf]: The overall problem here is, that `file_mod_item` extends `mod_item` but the former doesn't have anything like a `mod_body`. I decided to unify their apis (so users do not have to differentiate between mod and file mod) by creating artificial `getItemList` method on `mod_item`.

File: src/main/kotlin/org/rust/lang/core/psi/impl/mixin/RustModItemImplMixin.kt [hunk 2]
```diff
@@ -18,7 +18,7 @@ abstract class RustModItemImplMixin : RustItemImpl
     constructor(stub: RustItemStub, nodeType: IStubElementType<*, *>) : super(stub, nodeType)
 
     override val declarations: Collection<RustDeclaringElement>
-        get() = itemList
+        get() = itemList.orEmpty()
```
Comment (jajakobyly) [inline src/main/kotlin/org/rust/lang/core/psi/impl/mixin/RustModItemImplMixin.kt]: We do not want this

File: src/main/kotlin/org/rust/lang/core/resolve/RustResolveEngine.kt [hunk 3]
```diff
@@ -291,7 +291,7 @@ private class Resolver {
         override var matched: RustNamedElement? = null
 
         override fun visitModItem(mod: RustModItem) {
-            seek(mod.itemList)
+            seek(mod.itemList.orEmpty())
```
Comment (jajakobyly) [inline src/main/kotlin/org/rust/lang/core/resolve/RustResolveEngine.kt]: Again

**Comment by matklad** [comment]
I am not happy about this `fileModItem` situation at all... It also causes https://github.com/intellij-rust/intellij-rust/issues/354.

Maybe we can make `RustFile` and `RustModItem` to extend some kind of `RustMod`?

**Comment by jajakobyly** [comment]
Seems reasonable

**Comment by matklad** [comment]
>Implement this here?

Definitely not here, I predict a massive fallout after this change, especially in the resolve area. It would be cool if you implement this, but I am not sure about the design, so expect quite a bit of back and forth. If you don't want to mess with it, I think I can fix this tomorrow. 

I'll be glad to accept `foreign_mod` part right away!

**Comment by jajakobyly** [comment]
I think you can take on this refactoring. Also the `mod_item` part works, this method will never return null; it's just GK which puts `@Nullable` there. I could replace `orEmpty()` with `!!` and there would be no harm.

**Comment by matklad** [comment]
Ok, I really hope that I have not underestimate the amount of work the refactoring :)

**Comment by alexeykudinkin** [comment]
Are there any reasons for the new PSI element (`mod_body`) introduction?

**Comment by matklad** [comment]
@alexeykudinkin yep: https://github.com/intellij-rust/intellij-rust/pull/350#issuecomment-213802679.

**Comment by alexeykudinkin** [comment]
@matklad no, it's bad idea to fix indenter and expand-selection by introduction of the new PSI node. 

Can hardly spot any other reason for it to be introduced.

**Comment by matklad** [comment]
@alexeykudinkin and what are the benefits of not introducing the PSI node? I don't feel that either way is significantly better, but simpler formatter is a nice to have.

**Comment by alexeykudinkin** [comment]
> @alexeykudinkin and what are the benefits of not introducing the PSI node?

Mitigating introduction of the useless abstraction is the greatest benefit. The problem is with formatter & expand-selection not with the node.

**Comment by matklad** [comment]
Sounds reasonable, @jajakobyly what do you think?

**Comment by matklad** [comment]
I've looked at Java and Kotlin implementations, and the interesting patterns is that `{ stuff }` has its own PsiNode iff it is optional. 

That is, in Java `PsiClass` always has `{ }`, so there is no `PsiClass`. In Kotlin `{}` is optional, so there are `KtClass` and `KtClassBody`. 

I think this is a reasonable approach, based on this, in Rust there should be separate body elements for structs, enums and methods. Mods should not have a separate `ModBody` because module without a body is a `mod_decl`.

**Comment by jajakobyly** [comment]
Your arguments sound reasonable for me too, though I prefer introducing separate PSI nodes for such things as `mod_body` as I perceive them as hmm *distinguishable* entities.

Formatter can be easily adopted to the lack of `mod_body`, expand selection unfortunately is another story, as I cannot see any abilities to configure it ([ExpandSelectionAction](https://upsource.jetbrains.com/idea-ce/file/idea-ce-52c7416b07ef88943ab4cccebc34b4c3a304614a/plugins/ui-designer/src/com/intellij/uiDesigner/actions/ExpandSelectionAction.java?nav=0:0:focused)) (and it would be actually nice to have an ability to expand selection to the `{ ... }` range, because of already mentioned `mod_decl`. Though I understand that this is very minor feature.

So should I narrow this PR to `foreign_mod_body`, or close it completely?

**Comment by matklad** [comment]
>So should I narrow this PR to foreign_mod_body, or close it completely?

`foreign_mod_body` is also mandatory, so let's close this.

**Pull request closed without merging**